[Controls] Fix error thrown on numeric options list

Heenawter opened the pull request:
## Summary

This PR makes the suggestions returned from the options list route more type safe by ensuring that the defined return type (which was previously just `string`) aligns with the actual types returned from Elasticsearch (`string | number`). It also makes sure that selections are stored via this new `string | number` type, since this has been their real type all along despite being defined as just `string`. Since the EUI component requires `key` to be a string and this is how we determine what the user clicked on, we need to convert the `key` to a number when necessary on selection.


This resolves the error that was getting thrown for numeric options list controls, which was happening because we were returning **numbers** from the suggestions route rather than strings and some recent changes to the `EuiSelectable` component require strings:

**Before:**

https://github.com/user-attachments/assets/0e723e2f-e8f0-4466-b857-8164088cd1e7

**After**

https://github.com/user-attachments/assets/d9b138b9-de27-4e14-8c85-0ce4bfde16ce


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)


<!--ONMERGE {"backportTargets":["8.15"]} ONMERGE-->

Heenawter committed:
Ensure suggestions are always strings

Heenawter commented:
/ci

Heenawter commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

Heenawter commented:
/ci

Heenawter committed:
Clean up unnecessary dependency

Heenawter committed:
Add tests

Heenawter commented:
/ci

elasticmachine commented:
Pinging @elastic/kibana-presentation (Team:Presentation)

nreese commented on src/plugins/controls/server/options_list/suggestion_queries/options_list_search_suggestions.ts:
```diff
@@ -118,7 +118,7 @@ const suggestionAggSubtypes: { [key: string]: OptionsListSuggestionAggregationBu
       const suggestions = get(rawEsResult, `${basePath}.suggestions.buckets`)?.reduce(
         (acc: OptionsListSuggestions, suggestion: EsBucket) => {
-          acc.push({ value: suggestion.key, docCount: suggestion.doc_count });
+          acc.push({ value: String(suggestion.key), docCount: suggestion.doc_count });
```
Can you add a test case src/plugins/controls/server/options_list/suggestion_queries/options_list_search_suggestions.test.ts with a response that contains numeric key values.

Heenawter replied:
Done in d44dc552edaebe827101feeafd146a5b3826761c

Heenawter replied:
Oops, realized there was some duplication happening - cleaned up in 4dc0515aac8927623b46c095fce6587832423e20

nreese commented on test/functional/apps/dashboard_elements/controls/options_list/options_list_suggestions.ts:
```diff
@@ -119,6 +119,27 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
         await testSubjects.missingOrFail('dashboardUnsavedChangesBadge');
       });
+
+      it('can sort numeric options lists suggestions', async () => {
```
Functional tests are more expensive (meaning time) to maintain and run. I think this case would be better covered by adding a numeric test case to src/plugins/controls/server/options_list/suggestion_queries/options_list_search_suggestions.test.ts.

Heenawter replied:
@ThomThomson suggested having at least one smoke test to make sure that the options list popover can be opened without errors, and that is why I added this sorting functional test - I am personally in favour of keeping at least one of these two new tests in for that reason.

nreese commented on src/plugins/controls/public/hooks/use_field_formatter.ts:
```diff
@@ -29,10 +31,13 @@ export const useFieldFormatter = ({
+        dataView?.getFormatterForField(fieldSpec).getConverterFor('text') ??
+        ((toFormat: string) => toFormat);
+      setFieldFormatter(() =>
+        fieldSpec.type === 'date' // date fields must be passed in as a number for the formatter to work
```
Instead of doing the string conversion on the server and then having to have custom handling for edge cases in the client, what about keeping the server code as-is and then converting formatted values to string on the client?

Heenawter replied:
I thought about this, but it felt messier to have the server return an unknown type? Right now, our server-side code is typed so that it is **supposed** to return only strings, but we never did any casting to ensure this - so it was returning numbers for both numeric and date fields despite what the typing suggested. To keep this type safe, it felt safer to me to do the casting on the server side because of this - otherwise, I would have to go around and change the typing to "unknown" everywhere 🤷 What do you think?

nreese replied:
Since the data may be needed in its original form, we should not cast until the last minute. I would type the `value` as `string | number` and then cast to string before passing the value or formatted value to options list.

@ThomThomson What do you think?

Heenawter replied:
That's fair - I actually originally had `string | number` but I bounced to this idea instead because it felt cleaner to have the casting at a single point (server side) rather than all over the client-side code. I'm down to make that change if @ThomThomson agrees, though!

ThomThomson replied:
I'm not a huge fan of this extra layer of cast before the formatter. I'd recommend having the serverside code return `string | number` - which will likely cause a pretty large number of type failures clientside - then have those resolved by some `typeof` check before they end up in the formatter.

IMO the choice is between...

**A:** ES `string | number` -> server `string` -> client uses string or casts to number for dates
**B:** ES `string | number` -> server `string | number` -> client narrows to string or number

B seems more straightforward to me because it maintains the real types further along... 

> all over the client-side code

What other places do you anticipate needing this type narrowing? Maybe this is something that only the Options List Embeddable code needs to handle when the server returns results?

Heenawter replied:
Resolved - see https://github.com/elastic/kibana/pull/188789#issuecomment-2246338726 :+1:

Heenawter committed:
Add numeric field search tests

Heenawter committed:
Merge branch 'main' into fix-range-slider-value_2024-07-19

Heenawter committed:
Do some test cleanup

Heenawter committed:
Make types consistent

Heenawter committed:
Define selections as type

Heenawter committed:
Better error handling

Heenawter committed:
Fix types of query tests

Heenawter committed:
Merge branch 'main' into fix-range-slider-value_2024-07-19

Heenawter commented on src/plugins/controls/public/options_list/embeddable/options_list_embeddable.tsx:
```diff
@@ -437,14 +438,13 @@ export class OptionsListEmbeddable
       existsSelected,
       exclude,
-      selectedOptions: validSelections,
+      selectedOptions,
```
We aren't ignoring filters anymore, so this never should have been `validSelections`

Heenawter committed:
Another small test fix + cleanup

Heenawter commented:
@nreese @ThomThomson Ooof, that was a pain 😮‍💨 But I think I've cleaned up all the types!! Now, **anything** related to suggestions and/or selections is typed appropriately as `string | number` - but since `EuiSelectable` expects strings for both the `key` and `value` props, I've added casting as necessary on the client side. LMK what you think 🙇

Heenawter committed:
Use helper method for validation query

Heenawter committed:
Final (?) cleanup

Heenawter committed:
Fix missed import

nreese commented on src/plugins/controls/common/options_list/options_list_selections.ts:
```diff
@@ -0,0 +1,16 @@
+
+export type OptionsListSelection = string | number;
+
+export const getSelectionAsFieldType = (field: FieldSpec, key: string): OptionsListSelection => {
```
Is this function needed? How about just calling `String(key)` where key needs to be a string

Heenawter replied:
It is, yes - we need to be able to convert the key to a number where necessary.

Heenawter committed:
Fix failing tests + fix types

Heenawter commented on src/plugins/controls/common/options_list/options_list_selections.ts:
```diff
@@ -0,0 +1,16 @@
+
+import { FieldSpec } from '@kbn/data-views-plugin/common';
+
+export type OptionsListSelection = string | number;
```
That's fair! `OptionsListOption` is a little... funky sounding 😆 But maybe it's better?

Heenawter replied:
Renamed in a7524210ec3bb5fc434765452d42efd2d522055e :+1:

Heenawter replied:
Actually, scratch that - relooking at things and it **is** something the user has picked - always. We are only using this type **for selection based things**, so I think the original name is accurate. Reverting the previous rename.

Heenawter committed:
Rename `OptionsListSelections`

Heenawter committed:
Revert "Rename `OptionsListSelections`"

This reverts commit a7524210ec3bb5fc434765452d42efd2d522055e.

Heenawter committed:
Merge branch 'main' into fix-range-slider-value_2024-07-19

Heenawter committed:
Merge branch 'main' into fix-range-slider-value_2024-07-19

Heenawter committed:
Merge branch 'main' into fix-range-slider-value_2024-07-19

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:bundle:endpoint-management'

Heenawter committed:
Revert "[CI] Auto-commit changed files from 'yarn openapi:bundle:endpoint-management'"

This reverts commit 56dc66730ed315f5f0c8c9a856e2e63200cbbb4c.

Heenawter committed:
Merge branch 'main' into fix-range-slider-value_2024-07-19

nreese commented on src/plugins/controls/public/options_list/components/options_list_popover_invalid_selections.tsx:
```diff
@@ -91,8 +92,13 @@ export const OptionsListPopoverInvalidSelections = () => {
         listProps={{ onFocusBadge: false, isVirtualized: false }}
         onChange={(newSuggestions, _, changedOption) => {
+          if (!fieldSpec || !changedOption.key) {
+            throw new Error(OptionsListStrings.popover.getInvalidSelectionMessage());
```
Where is this error caught when thrown? I think you will need to handle the error in this component and maybe display the error. message in a toast.

Heenawter replied:
This actually shouldn't ever be hit - but it is technically being caught by the `EuiErrorBoundary`. I'm not sure if it's worth handling this more elegantly since we can handle this better with the new system? (considering again, this error was thrown as a type safeguard and should never happen)

Heenawter replied:
Switched to early return in c9672eedee5c99962f734984936a1a9da0001d6a

nreese replied:
This is what the rendering would look like if the error is thrown in the callback.

<img width="500" alt="Screenshot 2024-07-26 at 1 50 08 PM" src="https://github.com/user-attachments/assets/810fb13c-ae6c-44eb-9f80-aeb5604c8876">

At a minimum, the callback should not throw and console log a warning and return early. I don't think the argument of "this will never occur" holds water. 1) things change and the error may be triggered by future changes. 2) it provide a poor example that may be duplicated in other locations.

Heenawter replied:
Yeah that's fair - I think the error thrown was overkill :+1: Switched to early return in https://github.com/elastic/kibana/commit/c9672eedee5c99962f734984936a1a9da0001d6a

nreese commented on src/plugins/controls/public/options_list/components/options_list_popover_suggestions.tsx:
```diff
@@ -191,12 +198,19 @@ export const OptionsListPopoverSuggestions = ({
           emptyMessage={<OptionsListPopoverEmptyMessage showOnlySelected={showOnlySelected} />}
           onChange={(newSuggestions, _, changedOption) => {
-            const key = changedOption.key ?? changedOption.label;
+            if (!fieldSpec || !changedOption.key) {
```
Same here, callbacks can not throw because then the error is bubbled up to error boundary or top level react component

Heenawter replied:
See https://github.com/elastic/kibana/pull/188789#discussion_r1693507540 - I'm not convinced it's worth handling here, since this error should never actually be seen 🤷 I could technically just return early on that case rather than throwing an error - perhaps that is clearer?

Heenawter replied:
Switched to early return in c9672eedee5c99962f734984936a1a9da0001d6a

nreese replied:
How about a console.warn? That way if this ever occurs, there is still some log that we could use to debug the problem?

Heenawter replied:
Good call - added `console.warn` in 0edfa0fa1a0d5a85cae025d1d711ec17e523f7e5 :+1:

Heenawter committed:
Early return rather than thrown error

Heenawter committed:
Fix field formatter type

Heenawter committed:
Add `console.warn`

nreese reviewed (approved):
LGTM - thanks for fixing
code review, tested in chrome

logeekal reviewed (approved):
Desk tested from Investigations side. Looks great. Thanks.

Heenawter committed:
Merge branch 'main' into fix-range-slider-value_2024-07-19

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/189422">](https://github.com/elastic/kibana/pull/189422)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->

EricBrzezenski commented:
This issue also affects 7.17.x versions, Can this change be back-ported to the last release in the 7.17 branch?

Heenawter commented:
@EricBrzezenski The new controls were not available in `7.17.x` - they were released in `8.3.0`, and this fix targets those specifically. Do you perhaps mean that this error also impacts the old legacy controls? In that case, that feature is deprecated and will not be receiving any updates 🙇

EricBrzezenski commented:
Understood and Thank you @Heenawter for those details - I didn't realize there was a difference between controls.
